Implement control flow

SafelySwift opened the pull request:
This PR adds three new opcodes: "Jump", "Conditional jump if true" and "Go to".

"Jump": `0x70` -- offsets the program counter by the given amount
"Conditional jump if true": `0x71` -- offsets the program counter by the given amount if a condition is met
"Go to": `0x72` -- changes the program counter to the given offset

`reference.txt` was updated accordingly

# NOTE: For some reason updating `reference.txt` causes the build to fail?

Commit:
Add jump opcode

Commit:
Add cond jump "if true"

Commit:
Update and rename reference.txt to reference.md

Commit:
Rename reference.md to reference.txt

SafelySwift commented:
# For some reason updating `reference.txt` causes the build to fail?

Commit:
Add `goto` opcode

Commit:
Add `goto` docs

Commit:
Fix typo

kgscialdone commented on src/sherwood.cr:
```diff
@@ -95,6 +95,7 @@ class Sherwood
 
       # TODO: Variable Operations
       # TODO: Control Flow
+      when 0x70 then insp -= popType(SWInt, stack)
```
I'm not sure I really understand the purpose of this opcode. What advantage does it give over `goto` aside from relative positioning, and why is it `-=` instead of `+=`?

SafelySwift replied:
Idk why I did `-=` thats just me

kgscialdone commented on src/sherwood.cr:
```diff
@@ -95,6 +95,15 @@ class Sherwood
       # TODO: Control Flow
+      when 0x70 then insp -= popType(SWInt, stack)
+      when 0x71 then
+        if popType(Bool, stack) then {
```
Your code failing the test suite doesn't appear to have been from renaming `reference.txt`, but rather from this causing a compile error and subsequent test failures. Crystal's `if` syntax is this:
```crystal
if cond
  # stuff
else
  # stuff
end
```
Not this:
```crystal
if cond then {
  # stuff
} else {
  # stuff
}
```
It would probably be best if you ran the tests with `crystal spec` before committing =) Travis CI does run the tests for you, but it isn't very good about telling you exactly *where* they failed, and just gives you a big red X instead unless you have access to the logs. Running the tests from the command line will give you much better feedback on what needs to be fixed.

SafelySwift replied:
Idk how Crystal works so sorry, I can't. Ill have to rely on travis, but thats for the notice.

kgscialdone replied:
It's just a command line call to the compiler. The `crystal spec` command from the root directory of the project should run the tests without any fuss.

kgscialdone commented on src/sherwood.cr:
```diff
@@ -95,6 +95,15 @@ class Sherwood
+      when 0x71 then
+        if popType(Bool, stack) then {
+          popType(SWInt, stack)
+        } else {
```
This empty `else` is completely unnecessary.

SafelySwift replied:
Oh. 
I don't know how Crystal works, but could figure most stuff out. I looked at a tutorial which said it was required. I did not to take any chances 😄

kgscialdone commented on reference.txt:
```diff
@@ -80,4 +80,10 @@ SECTION: IO Operations
+= opcd name (params)
+==============================================================================================================
+  0x70 jump (o:int)         -- Changes the program counter by the given offset
+  0x71 cjit (o:int, c:bool) -- Changes the program counter by the given offset if the condition is met
```
This naming choice is a bit odd to me. Why `cjit` specifically, instead of something that more obviously refers to control flow for those who are perhaps unfamiliar with assembly slang?

SafelySwift replied:
Maybe jmpc for jump conditionally?

kgscialdone replied:
That would probably work.

kgscialdone reviewed (changes requested):
I appreciate the PR! Unfortunately, I'm not sure it aligns with my goals for Sherwood - primarily because I'm extremely hesitant to hand arbitrary control of the instruction pointer to users. I was planning to implement labeled `goto` instead, and restrict jumps to explicitly defined labels. Arbitrary `goto` could also become quite confusing to actually work with for people who are used to other VMs, as the instruction pointer isn't an arbitrary pointer into the bytecode source; rather, it's an array index into a pre-parsed list of operations (essentially chunking the bytecode into a pseudo-AST based on known opcode sizes, then operating off of that rather than the raw bytecode).

I'd also appreciate if you could take a look at `spec/opcode_spec.cr` and update it to add tests for your changes. You can look at the other examples there as well as the `opcode` macro defined in `spec/spec.cr` to get an idea of how to do so.

SafelySwift commented:
> I'm not sure it aligns with my goals for Sherwood

> I'm extremely hesitant to hand arbitrary control of the instruction pointer to users

Sorry! This is the only way I know how to implement control flow.
In order to resolve this, maybe hold a dictionary (idk what they are called in Crystal I just looked at your code :P) of labels and their insp vals. Then, when you use one, simply compile it as a insp value. This way you don't hand over the entire system to users, but allow for some flexibility.

swordglowsblue commented:
> In order to resolve this...

Most likely, my solution would be similar to this, but minus the compilation step. Once the file has been parsed into distinct `Opcode`s, it would be fairly easy to scan through for labels and build a mapping of them before execution, which would then be used as a jump table. In fact, that's the entire reason I pre-parse in the first place, though it does have some other benefits =)

Commit:
Fix if statement

Update according to [this discussion](https://github.com/swordglowsblue/sherwood/pull/2)

swordglowsblue commented:
Closed in favor of #3.
